lib/xnvme_be_vfio_mem.c

```diff
@@ -84,6 +84,53 @@ xnvme_be_vfio_buf_vtophys(const struct xnvme_dev *XNVME_UNUSED(dev), void *XNVME
 	errno = ENOSYS;
 	return -ENOSYS;
 }
+
+void *
+xnvme_be_vfio_buf_map(const struct xnvme_dev *dev, size_t nbytes, uint64_t *XNVME_UNUSED(phys), void *vaddr)
+{
+	struct xnvme_be_vfio_state *state = (void *)dev->be.state;
+	struct vfio_state *vfio = &state->ctrl->pci.vfio;
+  struct vfio_iommu_mapping *mapping;
+	
+	XNVME_DEBUG("xnvme_be_vfio_buf_map(%p, %p)", dev, vaddr);
+
+	mapping = vfio_iommu_find_mapping(&vfio->iommu, vaddr);
+	if (mapping) {
+		XNVME_DEBUG("Already mapped for %p\n", vaddr);
+		return vaddr;
+	}
+
+	nbytes = ALIGN_UP(nbytes, __VFN_PAGESIZE);
+
+	XNVME_DEBUG("vfio_map_vaddr(%p, %p, %ld, NULL)", vfio, vaddr, nbytes);
+	if (vfio_map_vaddr(vfio, vaddr, nbytes, NULL)) {
+		XNVME_DEBUG("FAILED: vfio_map_vaddr(): %s\n", strerror(errno));
+		return NULL;
+	}
+
+	return vaddr;
+}
+
+void
+xnvme_be_vfio_buf_unmap(const struct xnvme_dev *dev, void *buf)
```

Comment:
You think you want to be as generic as possible in the xNVMe API. Thus, you cannot rely on all implementations of mapping/unmapping buffer to be aware of the mapping length internally and should add a `length` parameter on this functions as well as allowing it to return an error code (mapping does not exist, unmapping failed, etc.).

Reply:
Thinking about this some more, I am not 100% convinced that my suggestions is sound.... It *is* as hazzle for the user to keep track of buffer mapping lengths, and you might want to just decide that if the backend does not implement it, then the xnvme "glue" has to take care of it. Not sure.